core/src/main/java/com/linecorp/armeria/common/logging/ContentPreviewerFactory.java

```diff
@@ -0,0 +1,151 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiPredicate;
+import java.util.function.Supplier;
+
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.RequestContext;
+
+@FunctionalInterface
+public interface ContentPreviewerFactory {
+    ContentPreviewer get(RequestContext ctx, HttpHeaders headers);
+
+    default ContentPreviewerFactory asImmutable() {
+        return this;
+    }
+
+    ContentPreviewerFactory DISABLED = (ctx, headers) -> ContentPreviewer.DISABLED;
+
+    /**
+     * TODO: AddJavadocs.
+     */
+    static ContentPreviewerFactory of(MediaType contentType, Supplier<ContentPreviewer> supplier) {
+        return of((ctx, headers) -> {
+            final MediaType type = headers.contentType();
+            return type != null && type.is(contentType);
+        }, supplier);
+    }
+
+    static ContentPreviewerFactory of(String contentType, Supplier<ContentPreviewer> supplier) {
+        return of(MediaType.parse(contentType), supplier);
+    }
+
+    /**
+     * TODO: AddJavadocs.
+     */
+    static ContentPreviewerFactory of(BiPredicate<RequestContext, HttpHeaders> predicate,
+                                      Supplier<ContentPreviewer> supplier) {
+        return (ctx, headers) -> {
+            if (predicate.test(ctx, headers)) {
+                return supplier.get();
+            }
+            return ContentPreviewer.DISABLED;
+        };
+    }
+
+    /**
+     * TODO: AddJavadocs.
+     */
+    static ContentPreviewerFactory of(ContentPreviewerFactory factory1, ContentPreviewerFactory factory2) {
+        if (factory1 == DISABLED) {
+            return factory2;
+        }
+        if (factory2 == DISABLED) {
+            return factory1;
+        }
+        if (factory1 instanceof ContentPreviewerFactoryCombinable) {
+            return ((ContentPreviewerFactoryCombinable) factory1).combine(factory2);
+        }
+        if (factory2 instanceof ContentPreviewerFactoryCombinable) {
+            return ((ContentPreviewerFactoryCombinable) factory2).combine(factory1);
+        }
+        return new CompositeContentPreviewerFactory(Arrays.asList(factory1, factory2));
+    }
+
+    /**
+     * TODO: AddJavadocs.
+     */
+    static ContentPreviewerFactory of(ContentPreviewerFactory... factories) {
+        final List<ContentPreviewerFactory> factoryList = new ArrayList<>();
+        ContentPreviewerFactory combined = DISABLED;
+        for (ContentPreviewerFactory factory : factories) {
+            if (factory == DISABLED) {
+                continue;
+            }
+            if (factory instanceof ContentPreviewerFactoryCombinable) {
+                combined = ((ContentPreviewerFactoryCombinable) factory).combine(combined);
+            } else if (combined instanceof ContentPreviewerFactoryCombinable) {
```

Comment:
Could just do `instanceof MappedContentPreviewerFactory` and access the entries directly?

Reply:
`factory` and `combined` could be also `CompositeContentPreviewerFactory`. So I think it would be too complicated if I do that?

Reply:
Ah, I see. Then, how about removing `Combinable` only?

Reply:
Sounds good to me.